test/thread/src/join_detach.d

```diff
@@ -0,0 +1,20 @@
+import core.thread;
+import core.sync.semaphore;
+
+__gshared Semaphore sem;
+
+void thread_main ()
+{
+    sem.notify();
+}
+
+void main()
+{
+    auto th = new Thread(&thread_main);
+    sem = new Semaphore();
+    th.start();
+    sem.wait();
+    while (th.isRunning()) {}
+    destroy(th); // force detach
```

Comment:
This triggers a spurious CI failure, e.g.:
```
*** Error in `./generated/linux/release/32/join_detach': double free or corruption (fasttop): 0xf6b004b8 ***
======= Backtrace: =========
/lib/i386-linux-gnu/libc.so.6(+0x67377)[0xf7612377]
/lib/i386-linux-gnu/libc.so.6(+0x6d2f7)[0xf76182f7]
/lib/i386-linux-gnu/libc.so.6(+0x6dc31)[0xf7618c31]
./generated/linux/release/32/join_detach(_D4core8internal9container6common8xreallocFNbNiPvkZQe+0x1e)[0x807d17e]
./generated/linux/release/32/join_detach(_D2rt5tlsgc7destroyFNbNiPvZv+0x8c)[0x807a93c]
./generated/linux/release/32/join_detach(_D4core6thread8osthread6Thread6__dtorMFNbNiZv+0x19)[0x8077f21]
./generated/linux/release/32/join_detach(rt_finalize2+0x58)[0x807fe88]
./generated/linux/release/32/join_detach(rt_finalize+0x1e)[0x8079d4e]
./generated/linux/release/32/join_detach(_D6object__T7destroyVbi1TC4core6thread8osthread6ThreadZQBrFNbQBkZv+0x1b)[0x807797f]
./generated/linux/release/32/join_detach(_Dmain+0x74)[0x8077928]
./generated/linux/release/32/join_detach(_D2rt6dmain212_d_run_main2UAAakPUQgZiZ6runAllMFZ9__lambda2MFZv+0x21)[0x8078f9d]
./generated/linux/release/32/join_detach(_D2rt6dmain212_d_run_main2UAAakPUQgZiZ7tryExecMFMDFZvZv+0x15)[0x8078e39]
./generated/linux/release/32/join_detach(_D2rt6dmain212_d_run_main2UAAakPUQgZiZ6runAllMFZv+0x8a)[0x8078f1a]
./generated/linux/release/32/join_detach(_D2rt6dmain212_d_run_main2UAAakPUQgZiZ7tryExecMFMDFZvZv+0x15)[0x8078e39]
./generated/linux/release/32/join_detach(_d_run_main2+0x193)[0x8078dc3]
./generated/linux/release/32/join_detach(_d_run_main+0x95)[0x8078c15]
./generated/linux/release/32/join_detach(main+0x1f)[0x807795f]
/lib/i386-linux-gnu/libc.so.6(__libc_start_main+0xf7)[0xf75c3637]
./generated/linux/release/32/join_detach[0x80777c1]
```

Reply:
This seems to be a race condition between the main and spawned thread both destroying the TLS. The current implementation has some measures to prevent double destruction, but those are not thread safe AFAICT.
Hence the explicit note [in the docs](https://dlang.org/phobos/core_thread_osthread.html#.Thread):

> [...] As thread manipulation is a required facility for garbage collection, all user threads should derive from this class, and instances of this class should never be explicitly deleted. [...]